operators/endpointmetrics/pkg/hypershift/hypershift.go

```diff
@@ -0,0 +1,276 @@
+	// Get the hypershift's etcd service monitor to replicate some of its settings
+	hypershiftEtcdSM := &promv1.ServiceMonitor{}
+	if err := c.Get(ctx, types.NamespacedName{Namespace: namespace, Name: "etcd"}, hypershiftEtcdSM); err != nil {
+		return nil, fmt.Errorf("failed to get hypershift's etcd ServiceMonitor: %w", err)
+	}
+
+	if len(hypershiftEtcdSM.Spec.Endpoints) != 1 {
+		return nil, fmt.Errorf("hypershift's etcd ServiceMonitor has more than one endpoint") // safe check
+	}
+
+	originalEndpoint := hypershiftEtcdSM.Spec.Endpoints[0]
+
+	return &promv1.ServiceMonitor{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      AcmEtcdSmName,
+			Namespace: namespace,
+		},
+		Spec: promv1.ServiceMonitorSpec{
+			Endpoints: []promv1.Endpoint{
+				{
+					Scheme:            "https",
+					Interval:          "15s",
+					Port:              originalEndpoint.Port,
+					BearerTokenSecret: corev1.SecretKeySelector{},
+					TLSConfig:         originalEndpoint.TLSConfig,
+					MetricRelabelConfigs: []*promv1.RelabelConfig{
+						{
+							SourceLabels: []string{"__name__"},
+							Action:       "keep",
+							Regex: "(etcd_server_has_leader|etcd_disk_wal_fsync_duration_seconds_bucket|" +
+								"etcd_mvcc_db_total_size_in_bytes|etcd_network_peer_round_trip_time_seconds_bucket|" +
+								"etcd_mvcc_db_total_size_in_use_in_bytes|" +
+								"etcd_disk_backend_commit_duration_seconds_bucket|" +
+								"etcd_server_leader_changes_seen_total)",
+						},
+						{
+							TargetLabel: "_id",
+							Action:      "replace",
+							Replacement: clusterID,
+						},
+						{
+							TargetLabel: "cluster_id",
+							Action:      "replace",
+							Replacement: clusterID,
+						},
+						{
+							TargetLabel: "cluster",
+							Action:      "replace",
+							Replacement: clusterName,
+						},
+					},
+					RelabelConfigs: []*promv1.RelabelConfig{
+						{
+							TargetLabel: "_id",
+							Action:      "replace",
+							Replacement: clusterID,
+						},
+						{
+							TargetLabel: "job",
+							Action:      "replace",
+							Replacement: "etcd",
+						},
+					},
+				},
+			},
+			Selector:          hypershiftEtcdSM.Spec.Selector,
+			NamespaceSelector: hypershiftEtcdSM.Spec.NamespaceSelector,
+		},
+	}, nil
+}
+
+func getKubeServiceMonitor(ctx context.Context, c client.Client, namespace, clusterID, clusterName string) (*promv1.ServiceMonitor, error) {
+	// Get the hypershift's api-server service monitor and replicate some of its settings
+	hypershiftKubeSM := &promv1.ServiceMonitor{}
+	if err := c.Get(ctx, types.NamespacedName{Namespace: namespace, Name: ApiServerSmName}, hypershiftKubeSM); err != nil {
+		return nil, fmt.Errorf("failed to get hypershift's kube-apiserver ServiceMonitor: %w", err)
+	}
+
+	if len(hypershiftKubeSM.Spec.Endpoints) != 1 {
+		return nil, fmt.Errorf("hypershift's kube-apiserver ServiceMonitor has more than one endpoint") // safe check
+	}
```

Comment:
What if it has zero endpoints? Same applies to other functions in this file that do a similar check.

Reply:
I mean, the `if` will do its job, but the log message will be wrong.

Reply:
indeed, changing that